compression_verifier.go

```diff
@@ -0,0 +1,225 @@
+package ghostferry
+
+import (
+	"crypto/md5"
+	"database/sql"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/golang/snappy"
+	"github.com/siddontang/go-mysql/schema"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// CompressionSnappy is used to identify Snappy (https://google.github.io/snappy/) compressed column data
+	CompressionSnappy = "SNAPPY"
+)
+
+// UnsupportedCompressionError is used to identify errors resulting
+// from attempting to decompress unsupported algorithms
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides support for verifying the payload of compressed columns that
+// may have different hashes for the same data by first decompressing the compressed
+// data before fingerprinting
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	// supportedAlgorithms provide O(1) lookup to check if a configured algorithm is supported
+	supportedAlgorithms map[string]struct{}
+
+	// tableColumnCompressions is represented as table -> column -> compression_algorithm
+	tableColumnCompressions map[string]map[string]string
```

Comment:
I think using a simple type annotation here would greatly improve readability. Something like:

```
type ColumnCompressionConfig map[string]string // map of columns to decompressors. e.g. { "body": "SNAPPY" }
```

you'd then have:

```
tableColumnCompressions map[string]ColumnCompressionConfig
```

which is a hash of `table -> ColumnCompressionConfig` which is readily comprehensible

Reply:
Added two types, one being exported (`TableColumnCompressionConfig`). Do you think this is more clear?